samples/nrf9160/serial_lte_modem/src/slm_at_icmp.c

```diff
@@ -0,0 +1,516 @@
+			hcs++;
+		}
+	}
+
+	ptr16 = (const u16_t *)ptr32;
+
+	if (len & 2) {
+		u16_t s = *ptr16++;
+
+		hcs += s;
+		if (hcs < s) {
+			hcs++;
+		}
+	}
+
+	if (len & 1) {
+		const u8_t *ptr8 = (const u8_t *)ptr16;
+		u8_t s = *ptr8;
+
+		hcs += s;
+		if (hcs < s) {
+			hcs++;
+		}
+	}
+
+	while (hcs > 0xFFFF) {
+		hcs = (hcs & 0xFFFF) + (hcs >> 16);
+	}
+
+	return ~hcs;    /* One's complement */
+}
+
+/* Taken from ipv4.c */
+static void calc_ics(u8_t *buffer, int len, int hcs_pos)
+{
+	u16_t *ptr_hcs = (u16_t *)(buffer + hcs_pos);
+	*ptr_hcs = 0;   /* Clear checksum before calculation */
+	u16_t hcs;
+
+	hcs = check_ics(buffer, len);
+	*ptr_hcs = hcs;
+}
+
+static u32_t send_ping_wait_reply(void)
+{
+	static u8_t seqnr;
+	u16_t total_length;
+	u8_t ip_buf[NET_IPV4_MTU];
+	u8_t *data = NULL;
+	u8_t rep = 0;
+	static s64_t start_t, delta_t;
+	const u8_t header_len = 20;
+	int pllen, len;
+	const u16_t icmp_hdr_len = 8;
+	struct sockaddr_in *sa;
+	struct nrf_pollfd fds[1];
+	int fd;
+	int ret;
+	int hcs;
+	int plseqnr;
+
+	/* Generate IPv4 ICMP EchoReq */
+	total_length = argv.len + header_len + icmp_hdr_len;
+	memset(ip_buf, 0x00, header_len);
+
+	/* IPv4 header */
+	ip_buf[0] = (4 << 4) + (header_len / 4); /* Version & header length */
+	ip_buf[1] = 0x00;                        /* Type of service */
+	ip_buf[2] = total_length >> 8;           /* Total length */
+	ip_buf[3] = total_length & 0xFF;         /* Total length */
+	ip_buf[4] = 0x00;                        /* Identification */
+	ip_buf[5] = 0x00;                        /* Identification */
+	ip_buf[6] = 0x00;                        /* Flags & fragment offset */
+	ip_buf[7] = 0x00;                        /* Flags & fragment offset */
+	ip_buf[8] = 64;                          /* TTL */
+	ip_buf[9] = ICMP;                        /* Protocol */
+	/* ip_buf[10..11] = ICS, calculated later */
+
+	sa = (struct sockaddr_in *)argv.src->ai_addr;
+	setip(ip_buf+12, sa->sin_addr.s_addr);     /* Source */
+	sa = (struct sockaddr_in *)argv.dest->ai_addr;
+	setip(ip_buf+16, sa->sin_addr.s_addr);     /* Destination */
+
+	calc_ics(ip_buf, header_len, 10);
+
+	/* ICMP header */
+	data = ip_buf + header_len;
+	data[0] = ICMP_ECHO_REQ;                 /* Type (echo req) */
+	data[1] = 0x00;                          /* Code */
+	/* data[2..3] = checksum, calculated later */
+	data[4] = 0x00;                         /* Identifier */
+	data[5] = 0x00;                         /* Identifier */
+	data[6] = seqnr >> 8;                   /* seqnr */
+	data[7] = ++seqnr;                      /* seqr */
+
+	/* Payload */
+	for (int i = 8; i < total_length - header_len; i++) {
+		data[i] = (i + seqnr) % 10 + '0';
+	}
+
+	/* ICMP CRC */
+	calc_ics(data, total_length - header_len, 2);
+
+	rep = ICMP_ECHO_REP;
+
+	/* Send the ping */
+	do {
+		errno = 0;
+		delta_t = 0;
+		start_t = k_uptime_get();
+
+		fd = nrf_socket(NRF_AF_PACKET, NRF_SOCK_RAW, 0);
+		if (fd < 0) {
+			LOG_ERR("socket() failed: (%d)", -errno);
+			break;
```

Comment:
Perhaps we could just return from the function here, instead of indenting this whole section. There's no other `break` in the loop (all other exit points are handled by `goto`).

Reply:
Yes, updated